### ISSUE TITLE
Feature Request:  recover from failed pull without restarting

### ISSUE THREAD

**Issue by marcpawl** [containers/podman#7497]
/kind feature

**Description**
When pulling down a an image large layer (>1GB) through Artifactory the download fails often towards the end.  This seems to be a known issue with Docker also, and has lots of questions posted.

When you pull the same image the download of the layer starts all over again, and the same problem occurs, so it may take many attempts to get the image. (And of course consume incredible amounts of bandwidth.)

It would be good to have the partial layer saved, and when the pull is attempted again for the download to resume from where it left off.  (Like most web browsers.)

**Steps to reproduce the issue:**

1.  Create image with very large layers and store in repository

2.  Pull image over slow and possibly unreliable network.

3. Have pull fail when layer is close to complete.

**Describe the results you received:**
4. Pull image again, download for layer restarts from zero.

**Describe the results you expected:**
4. Pull image again, download for layer restarts from point of failure.

**Additional information you deem important (e.g. issue happens only occasionally):**
Happens a lot

**Output of `podman version`:**

podman version 2.0.5

**Output of `podman info --debug`:**

```
 podman info --debug
host:
  arch: amd64
  buildahVersion: 1.15.1
  cgroupVersion: v2
  conmon:
    package: conmon-2.0.19-1.fc32.x86_64
    path: /usr/bin/conmon
    version: 'conmon version 2.0.19, commit: 5dce9767526ed27f177a8fa3f281889ad509fea7'
  cpus: 4
  distribution:
    distribution: fedora
    version: "32"
  eventLogger: file
  hostname: dell
  idMappings:
    gidmap:
    - container_id: 0
      host_id: 1000
      size: 1
    - container_id: 1
      host_id: 100000
      size: 65536
    uidmap:
    - container_id: 0
      host_id: 1000
      size: 1
    - container_id: 1
      host_id: 100000
      size: 65536
  kernel: 5.7.17-200.fc32.x86_64
  linkmode: dynamic
  memFree: 11666780160
  memTotal: 16671268864
  ociRuntime:
    name: crun
    package: crun-0.14.1-3.fc32.x86_64
    path: /usr/bin/crun
    version: |-
      crun version 0.14.1
      commit: 598ea5e192ca12d4f6378217d3ab1415efeddefa
      spec: 1.0.0
      +SYSTEMD +SELINUX +APPARMOR +CAP +SECCOMP +EBPF +YAJL
  os: linux
  remoteSocket:
    path: /run/user/1000/podman/podman.sock
  rootless: true
  slirp4netns:
    executable: /usr/bin/slirp4netns
    package: slirp4netns-1.1.4-1.fc32.x86_64
    version: |-
      slirp4netns version 1.1.4
      commit: b66ffa8e262507e37fca689822d23430f3357fe8
      libslirp: 4.3.1
      SLIRP_CONFIG_VERSION_MAX: 2
  swapFree: 8405381120
  swapTotal: 8405381120
  uptime: 13h 6m 16.68s (Approximately 0.54 days)
registries:
  search:
  - registry.fedoraproject.org
  - registry.access.redhat.com
  - registry.centos.org
  - docker.io
store:
  configFile: /home/mpawlowsky/.config/containers/storage.conf
  containerStore:
    number: 0
    paused: 0
    running: 0
    stopped: 0
  graphDriverName: overlay
  graphOptions:
    overlay.mount_program:
      Executable: /usr/bin/fuse-overlayfs
      Package: fuse-overlayfs-1.1.2-1.fc32.x86_64
      Version: |-
        fusermount3 version: 3.9.1
        fuse-overlayfs: version 1.1.0
        FUSE library version 3.9.1
        using FUSE kernel interface version 7.31
  graphRoot: /home/mpawlowsky/.local/share/containers/storage
  graphStatus:
    Backing Filesystem: xfs
    Native Overlay Diff: "false"
    Supports d_type: "true"
    Using metacopy: "false"
  imageStore:
    number: 6
  runRoot: /run/user/1000/containers
  volumePath: /home/mpawlowsky/.local/share/containers/storage/volumes
version:
  APIVersion: 1
  Built: 0
  BuiltTime: Wed Dec 31 19:00:00 1969
  GitCommit: ""
  GoVersion: go1.14.6
  OsArch: linux/amd64
  Version: 2.0.5

```

**Package info (e.g. output of `rpm -q podman` or `apt list podman`):**

```
podman-2.0.5-1.fc32.x86_64
```

**Have you tested with the latest version of Podman and have you checked the Podman Troubleshooting Guide?**

No

**Additional environment details (AWS, VirtualBox, physical, etc.):**

fedora 32, brand new install with updates applied.  Using openconnect for VPN to repo.

[mpawlowsky@dell nb99999999.9999]$ /usr/bin/podman pull XXX/YYY/src_trunk:2036
Trying to pull XXX/YYY/src_trunk:2036...
Getting image source signatures
Copying blob 840caab23da4 skipped: already exists  
Copying blob a30e5519f932 done  
Copying blob 4e379b10fe80 [==============>-----------------] 436.2MiB / 958.9MiB
  read tcp 1.220.29.191:45794->999.99.999.99:443: read: connection reset by peer
Error: unable to pull XXX/YYY/src_trunk:2036: Error writing blob: error storing blob to file "/var/tmp/storage207531848/2": read tcp 1.220.29.191:45794->999.99.999.99:443: read: connection reset by peer

[mpawlowsky@dell nb99999999.9999]$ find /var/tmp/storage207531848
find: ‘/var/tmp/storage207531848’: No such file or directory
[mpawlowsky@dell nb99999999.9999]$

**Comment by marcpawl** [comment]
Workaround for docker?
https://forums.docker.com/t/docker-pull-results-in-request-canceled-while-waiting-for-connection-client-timeout-exceeded-while-awaiting-headers/73064/7

**Comment by rhatdan** [comment]
@mtrmac PTAL

**Comment by mtrmac** [comment]
Taking advantage of _successfully pulled_ layers on future pulls might happen as a side effect of parallel layer commit Valentin was working on some time ago.

Restarting an _individual layer pull_ with a range request is problematic with the current architecture, where compression/decompression/encryption/decryption happens on the fly, so there may not be any obvious restart point. It seems feasible in principle for the pull path (where we seem to record the original blob — currently at least), but the complexity and API implications are not at all trivial.

**Comment by rhatdan** [comment]
@vrothberg FYI

**Comment by vrothberg** [comment]
@rhatdan, let's discuss in the next planning when to tackle it. We may well get this in for v3.0. Once Podman v2.1 is out, we have some wiggle room for bigger changes.

**Comment by rhatdan** [comment]
I agree this seems like a good 3.0 feature.

**Comment by marcpawl** [comment]
If you implement this, a flag to indicate that partial layers are being cached, and therefore worst performance/space, would be acceptable. e.g. --cache-layers, or --cache-layers-dir=<dir> or ....  

I was thinking you could do something like pull the compressed/encrypted file and store it all on disk until it is all copied.  And then use the disk copy as the source instead of the network.  (Wild idea not knowing anything about your code, so the idea might be completely wrong.)

**Comment by rhatdan** [comment]
@vrothberg I think we should move this up the priority list after shortnames is complete.

**Comment by vrothberg** [comment]
> @vrothberg I think we should move this up the priority list after shortnames is complete.

Before or after image volumes?

**Comment by rhatdan** [comment]
Probably after. :^(

**Comment by rhatdan** [comment]
@vrothberg do you have time to look at this?

**Comment by vrothberg** [comment]
> @vrothberg do you have time to look at this?

This would be a by-product of reworking the commit-logic in c/image.  It's a larger work item but it's up to Jedi council if I have time for it :)

**Comment by rhatdan** [comment]
This did not make it into 3.0 release.

**Comment by vrothberg** [comment]
@rhatdan I think we should tackle this item after bug week(s). Once this feature here is addressed, we can tackle the image package.

**Comment by rhatdan** [comment]
SGTM

**Comment by rhatdan** [comment]
@vrothberg Another on your list to improve pulls.

**Comment by vrothberg** [comment]
ACK.  That will actually be fixed with the "commit as early ass possible" item I am currently working on.

**Comment by vrothberg** [comment]
This has been released with containers/image v5.11.0 which has been merged into Podman :)

**Comment by alien999999999** [comment]
I have podman 3.0.0, where can I test this? I have a timeout due to a big file (can i just increase the timeout) and it retries download everytime, i get to 1.7GB (half the blob) and it retries everytime, eventually resulting in a /tmp out of disk space

**Comment by vrothberg** [comment]
3.0 is probably too old.  The new behavior shipped with Podman v3.2 but I suggest directly updating to 4.3 if possible.
